src/sfizz/ADSREnvelope.cpp

```diff
@@ -100,110 +99,92 @@ Type ADSREnvelope<Type>::getNextValue() noexcept
-        remainingSamples -= length;
-        attack -= length;
-        if (remainingSamples == 0)
+        switch (currentState) {
+        case State::Delay:
+            while (count < size && delay-- > 0) {
+                currentValue = start;
+                output[count++] = currentValue;
+            }
+            if (delay <= 0)
+                currentState = State::Attack;
             break;
-
-        currentValue = peak;
-        currentState = State::Hold;
-        // fallthrough
-    case State::Hold:
-        length = min(remainingSamples, hold);
-        fill<Type>(output, currentValue);
-        output.remove_prefix(length);
-        remainingSamples -= length;
-        hold -= length;
-        if (remainingSamples == 0)
+        case State::Attack:
+            while (count < size && (currentValue += peak * attackStep) < peak)
+                output[count++] = currentValue;
+            if (currentValue >= peak) {
+                currentValue = peak;
+                currentState = State::Hold;
+            }
             break;
-
-        step = std::exp(std::log(sustain + config::virtuallyZero) / (decay > 0 ? decay : 1));
-        currentState = State::Decay;
-        // fallthrough
-    case State::Decay:
-        length = min(remainingSamples, decay);
-        currentValue = multiplicativeRamp<Type>(output, currentValue, step);
-        output.remove_prefix(length);
-        remainingSamples -= length;
-        decay -= length;
-        if (remainingSamples == 0)
+        case State::Hold:
+            while (count < size && hold-- > 0)
+                output[count++] = currentValue;
+            if (hold <= 0)
+                currentState = State::Decay;
             break;
-
-        currentValue = sustain;
-        currentState = State::Sustain;
-        // fallthrough
-    case State::Sustain:
-        if (freeRunning)
-            shouldRelease = true;
-        break;
-    case State::Release:
-        length = min(remainingSamples, release);
-        currentValue = multiplicativeRamp<Type>(output, currentValue, step);
-        output.remove_prefix(length);
-        remainingSamples -= length;
-        release -= length;
-        if (remainingSamples == 0)
+        case State::Decay:
+            while (count < size && (currentValue *= decayRate) > sustain)
+                output[count++] = currentValue;
+            if (currentValue <= sustain) {
+                currentValue = sustain;
+                currentState = State::Sustain;
+            }
+            break;
+        case State::Sustain:
+            if (!shouldRelease && freeRunning) {
+                shouldRelease = true;
+                break;
+            }
+            count = size;
+            currentValue = sustain;
+            sfz::fill(output.first(count), currentValue);
+            break;
+        case State::Release:
+            while (count < size && (currentValue *= releaseRate) > config::virtuallyZero)
```

Comment:
So you're saying this could be done earlier? I'm not clear on the "I don't hear anything" level to be honest :)

Reply:
Usually -60 dB, -80 dB sometimes